src/VirtoCommerce.CatalogPersonalizationModule.Data/Services/PersonalizationService.cs

```diff
@@ -29,55 +28,9 @@ public class PersonalizationService : ServiceBase, ITaggedItemService, ITaggedIt
             _taggedEntitiesServiceFactory = taggedEntitiesServiceFactory;
         }
 
-        public TaggedItem[] GetTaggedItemsByIds(string[] ids, string responseGroup = null)
-        {
-            var retVal = new List<TaggedItem>();
-            if (ids != null)
-            {
-                using (var repository = _repositoryFactory())
-                {
-                    repository.DisableChangesTracking();
-
-                    retVal = repository.GetTaggedItemsByIds(ids, responseGroup)
-                        .Select(x => x.ToModel(AbstractTypeFactory<TaggedItem>.TryCreateInstance()))
-                        .ToList();
-                }
-            }
-            return retVal.ToArray();
-        }
-
-        public void SaveTaggedItems(TaggedItem[] taggedItems)
-        {
-            var pkMap = new PrimaryKeyResolvingMap();
-            using (var repository = _repositoryFactory())
-            {
-                using (var changeTracker = GetChangeTracker(repository))
-                {
-                    var ids = taggedItems.Select(x => x.Id).Where(x => x != null).Distinct().ToArray();
-                    var alreadyExistEntities = repository.GetTaggedItemsByIds(ids, null).ToArray();
-                    foreach (var taggedItem in taggedItems)
-                    {
-                        var sourceEntity = AbstractTypeFactory<TaggedItemEntity>.TryCreateInstance().FromModel(taggedItem, pkMap);
-                        var targetEntity = alreadyExistEntities.FirstOrDefault(x => x.Id == taggedItem.Id);
-                        if (targetEntity != null)
-                        {
-                            changeTracker.Attach(targetEntity);
-                            sourceEntity.Patch(targetEntity);
-                        }
-                        else
-                        {
-                            repository.Add(sourceEntity);
-                        }
-                    }
-                }
-                CommitChanges(repository);
-                pkMap.ResolvePrimaryKeys();
-            }
-        }
-
-        public GenericSearchResult<TaggedItem> SearchTaggedItems(TaggedItemSearchCriteria criteria)
+        public async Task<GenericSearchResult<TaggedItem>> SearchTaggedItemsAsync(TaggedItemSearchCriteria criteria)
         {
-            var retVal = new GenericSearchResult<TaggedItem>();
+            var result = new GenericSearchResult<TaggedItem>();
```

Comment:
Please use caching for search and all Get operations. 
In save and delete you need to `PersonalizationCacheRegion.ExpireRegion()`
[Example](https://github.com/VirtoCommerce/vc-module-marketing/blob/release/3.0.0/src/VirtoCommerce.MarketingModule.Data/Services/CouponService.cs#L37)

Reply:
done